### DISCUSSION TITLE
Freeplane lists aren't

### DISCUSSION THREAD

**Discussion by timboxyz** [freeplane/freeplane#1984] [Help] [answered]
According to the Feeplane API many methods such as `node.getPathToRoot()` have a return type of `List` and in the Scripting API map link to the [relevant documentation](http://docs.groovy-lang.org/latest/html/groovy-jdk/java/util/List.html) for that type. However, their type is reported as `class org.freeplane.plugin.script.proxy.ProxyUtils$1` as shown by:-
```
def route = node.getPathToRoot()
node.createChild(route.class)
```
and they don't accept some of the standard List methods listed in the linked docs such as` pop()` For example:-

```
def route = node.getPathToRoot()
route.pop()
```
produces a "Error executing script: null" message

Am I missing something here? Groovy is still quite new to me so I am still getting used to it.

**Comment by i-plasm** [comment, marked as answer]
The freeplane list you are getting is a custom implementation of the standard java list interface, so it does not feature the groovy enhancements - such as the `pop()` method. You can turn it into a groovy-enhanced list writing `def route = [] + node.getPathToRoot()`.

**Reply by timboxyz** [reply]
> The freeplane list you are getting is a custom implementation of the standard java list interface, so it does not feature the groovy enhancements - such as the `pop()` method. You can turn it into a groovy-enhanced list writing `def route = [] + node.getPathToRoot()`.

Ahh, that is very useful to know. What is not useful, and to my mind an extraordinary decision is to link from the Scripting API map to documentation which doesn't accurately represent the Class linked from. Linking to the Java version would seem more sensible given what you say.

I will try your workaround whose logic I follow. I had tried `def route = node.getPathToRoot() as List` which I had expected to do what I wanted but it doesn't, for reasons I do not understand.

**Reply by dpolivaev** [reply]
Would it make sense to return the Groovy list from the API?

**Reply by timboxyz** [reply]
Currently, that is what it seems to do. See the link above under "relevant documentation" which is the link I got from the API map. It makes sense to point to whatever documentation accurately represents the supported methods and behaviour of the Class as implemented in FP which, in this case, would seem to be the Java docs.

**Reply by dpolivaev** [reply]
I am sorry, I got no answer to my question.

**Reply by timboxyz** [reply]
> I will try your workaround whose logic I follow. I had tried `def route = node.getPathToRoot() as List` which I had expected to do what I wanted but it doesn't, for reasons I do not understand.

But after some digging think I do. What does seem to work, and feels somewhat tidier code than
`def route = [] + node.getPathToRoot()`, is what I presume is the correct coercion of
`def route =  node.getPathToRoot() as ArrayList`

**Reply by timboxyz** [reply]
> I am sorry, I got no answer to my question.

In which case I must have misunderstood what you were asking, sorry.

If you are asking if FP returned lists should be Groovy lists then yes since the expectation is to use the capabilities of Groovy within scripts. Failing that consistency between actual behaviour and documentation is important and some kind of note to warn that FP lists do not support all Groovy methods with perhaps a hint on how to coerce them to be so would help.

Is that what you were asking about?
